File: apps/ios/Sources/Services/HubActivityService.swift
import Foundation

// MARK: - HubActivityState

/// Per-hub live activity state reflecting the latest relay events for that hub.
struct HubActivityState: Equatable {
    /// Whether the current user is currently on shift for this hub.
    var isOnShift: Bool = false
    /// Number of active (ringing or in-progress) calls for this hub.
    var activeCallCount: Int = 0
    /// Number of unread messages for this hub since last opened.
    var unreadMessageCount: Int = 0
    /// Number of unread/unactioned conversation assignments for this hub since last opened.
    var unreadConversationCount: Int = 0
}

// MARK: - HubActivityService

/// Tracks per-hub live activity state driven by `AttributedHubEvent` events from
/// `WebSocketService.attributedEvents`. Each hub's state is maintained independently,
/// allowing the UI to display per-hub badges and indicators in a multi-hub list.
///
/// Thread-safe via `NSLock`. State mutations are lock-protected; reads return a
/// value-type copy so callers never hold a reference into the internal map.
///
/// Wire into `AppState` by consuming `webSocketService.attributedEvents` in a
/// background task and calling `handle(_:)` for each event.
@Observable
final class HubActivityService {

    // MARK: - Private State

    private var states: [String: HubActivityState] = [:]
    private let stateLock = NSLock()

    // MARK: - Public API

    /// Returns the current activity state for the given hub ID.
    /// Returns a default zero state if no events have been received for that hub.
    func state(for hubId: String) -> HubActivityState {
        stateLock.lock()
        defer { stateLock.unlock() }
        return states[hubId] ?? HubActivityState()
    }

    /// Process a hub-attributed event and update the corresponding hub's state.
    func handle(_ attributed: AttributedHubEvent) {
        stateLock.lock()
        var s = states[attributed.hubId] ?? HubActivityState()

        switch attributed.event {
        case .callRing:
            s.activeCallCount += 1

        case .callAnswered:
            // A call being answered transitions it from ringing to in-progress.
            // Decrement ringing count; the call is now active but already counted.
            // Guard against going negative (e.g. on reconnect with missed ring event).
            s.activeCallCount = max(0, s.activeCallCount - 1)

        case .callEnded:
            s.activeCallCount = max(0, s.activeCallCount - 1)

        case .callUpdate:
            // Status update only — no count change.
            break

        case .shiftStarted:
            s.isOnShift = true

        case .shiftEnded:
            s.isOnShift = false

        case .shiftUpdate:
            // Generic shift update — no state change without explicit start/end signal.
            break

        case .messageNew:
            s.unreadMessageCount += 1

        case .messageStatus:
            // Delivery/read receipt — no unread count change.
            break

        case .conversationNew, .conversationAssigned:
            s.unreadConversationCount += 1

        case .conversationClosed:
            s.unreadConversationCount = max(0, s.unreadConversationCount - 1)

        case .noteCreated, .voicemailNew, .presenceSummary, .presenceDetail, .unknown:
            break
        }

        states[attributed.hubId] = s
        stateLock.unlock()
    }

    /// Mark a hub as opened, clearing its unread message and conversation counts.
    /// Call this when the user navigates into a hub's dashboard.
    func markHubOpened(_ hubId: String) {
        stateLock.lock()
        var s = states[hubId] ?? HubActivityState()
        s.unreadMessageCount = 0
        s.unreadConversationCount = 0
        states[hubId] = s
        stateLock.unlock()
    }
}

File: apps/android/app/src/main/java/org/llamenos/hotline/hub/HubActivityService.kt
package org.llamenos.hotline.hub

import org.llamenos.hotline.model.LlamenosEvent
import org.llamenos.hotline.service.AttributedHubEvent
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton

data class HubActivityState(
    val isOnShift: Boolean = false,
    val activeCallCount: Int = 0,
    val unreadMessageCount: Int = 0,
    val unreadConversationCount: Int = 0,
)

/**
 * Tracks per-hub activity state derived from attributed Nostr relay events.
 *
 * State is maintained in-memory as a [ConcurrentHashMap] keyed by hub ID.
 * Callers feed events via [handle]; UI reads current state via [state].
 * [markHubOpened] resets unread counters when the user navigates to a hub.
 */
@Singleton
class HubActivityService @Inject constructor() {

    private val states = ConcurrentHashMap<String, HubActivityState>()

    /** Returns the current [HubActivityState] for [hubId], or a default empty state. */
    fun state(hubId: String): HubActivityState = states.getOrDefault(hubId, HubActivityState())

    /**
     * Process an [AttributedHubEvent] and update the corresponding hub's state.
     *
     * Uses sealed-class pattern matching so all event types are handled exhaustively
     * at compile time. Unknown/irrelevant events leave the state unchanged.
     */
    fun handle(attributed: AttributedHubEvent<LlamenosEvent>) {
        states.compute(attributed.hubId) { _, current ->
            val s = current ?: HubActivityState()
            when (val event = attributed.event) {
                is LlamenosEvent.CallRing ->
                    s.copy(activeCallCount = s.activeCallCount + 1)
                is LlamenosEvent.CallAnswered ->
                    s.copy(activeCallCount = maxOf(0, s.activeCallCount - 1))
                is LlamenosEvent.CallEnded ->
                    s.copy(activeCallCount = maxOf(0, s.activeCallCount - 1))
                is LlamenosEvent.VoicemailNew ->
                    s.copy(activeCallCount = maxOf(0, s.activeCallCount - 1))
                is LlamenosEvent.ShiftUpdate -> when (event.status) {
                    "started" -> s.copy(isOnShift = true)
                    "ended" -> s.copy(isOnShift = false)
                    else -> s
                }
                is LlamenosEvent.MessageNew ->
                    s.copy(unreadMessageCount = s.unreadMessageCount + 1)
                is LlamenosEvent.ConversationAssigned ->
                    s.copy(unreadConversationCount = s.unreadConversationCount + 1)
                is LlamenosEvent.ConversationClosed ->
                    s.copy(unreadConversationCount = maxOf(0, s.unreadConversationCount - 1))
                is LlamenosEvent.CallUpdate,
                is LlamenosEvent.NoteCreated,
                is LlamenosEvent.ConversationNew,
                is LlamenosEvent.PresenceSummary,
                is LlamenosEvent.PresenceDetail,
                is LlamenosEvent.MessageStatus,
                is LlamenosEvent.Unknown -> s
            }
        }
    }

    /**
     * Reset unread message and conversation counts for [hubId].
     * Called when the user opens/navigates to a hub to mark its content as seen.
     */
    fun markHubOpened(hubId: String) {
        states.compute(hubId) { _, current ->
            (current ?: HubActivityState()).copy(
                unreadMessageCount = 0,
                unreadConversationCount = 0,
            )
        }
    }
}
